implementation/src/main/java/io/smallrye/mutiny/groups/UniOnPredicate.java

```diff
@@ -0,0 +1,375 @@
+     * its result, the result is discarded, and the original (non null) {@code item} is forwarded downstream. If the
+     * produced {@code Uni} fails, the failure is propagated downstream.
+     *
+     * @param action the callback, must not be {@code null}
+     * @return the new {@link Uni}
+     */
+    public Uni<T> call(Function<? super T, Uni<?>> action) {
+        Function<? super T, Uni<?>> actual = Infrastructure.decorate(nonNull(action, "action"));
+        return upstream.onItem().call(item -> {
+            if (predicate.test(item)) {
+                return actual.apply(item);
+            } else {
+                return Uni.createFrom().nullItem();
+            }
+        });
+    }
+
+    /**
+     * Produces a new {@link Uni} invoking the given {@code action} when the {@code item} matches the predicate. Note
+     * that if the received item does not match the predicate, the action is not executed.
+     * <p>
+     * Unlike {@link #invoke(Consumer)}, the passed function returns a {@link Uni}. When the produced {@code Uni} sends
+     * its result, the result is discarded, and the original (non null) {@code item} is forwarded downstream. If the
+     * produced {@code Uni} fails, the failure is propagated downstream.
+     *
+     * @param action the callback, must not be {@code null} and must not return {@code null}
+     * @return the new {@link Uni}
+     */
+    public Uni<T> call(Supplier<Uni<?>> action) {
+        Supplier<Uni<?>> actual = Infrastructure.decorate(nonNull(action, "action"));
+        return call(ignored -> actual.get());
+    }
+
+    /**
+     * Produces a new {@link Uni} invoking the given function when the current {@link Uni} fires the {@code item} event
+     * and its item matches the predicate. The function receives the item as parameter, and can transform it. The
+     * returned object is sent downstream as {@code item}.
+     * <p>
+     * If the item does not match the predicate, the mapper is not called and it produces a {@code null} item.
+     * <p>
+     * Please note that the original item not matching the predicate is <b>not</b> retained in that case since it might
+     * not be of the same type as the result of the mapper function. Therefore, to avoid having a pipeline with
+     * ambiguous type information, {@code null} is returned instead. If you want to have a different behavior for this
+     * case, please use {@link #transformElse(Function, Function)}.
+     * <p>
+     * For asynchronous composition, see {@link #transformToUni(Function)}.
+     *
+     * @param mapper the mapper function called if the item matches the predicate, must not be {@code null}
+     * @param <R> the type of Uni item
+     * @return the new {@link Uni}
+     */
+    public <R> Uni<R> transform(Function<? super T, ? extends R> mapper) {
+        return transformElse(mapper, item -> null);
```

Comment:
Also added an example to the JavaDoc of `transform`, `transformToUni` etc. to make this more clear.

Reply:
I'm currently revisiting this PR. I don't have much to say on the implementation, it's clean, tested and documented 👍 

I'm wondering if this is an addition we want to have, or if it's making the API surface bigger. I'm especially unsure of the `else` variants. I'd like to take some further time to think about it.

Reply:
@jponge Sure. I mainly included the `else` variants because I felt like the user would be too limited in what happens in that case, since the original `item` is not preserved. So without the `else` variants, the use cases where `when()` would actually be useful would be limited to those where you only want to do something for one case, but nothing in the other case. But I agree that it adds some methods to the API which aren't available anywhere else.

An alternative that I thought of might be to actually "split" the Uni into two new Unis, one for the case when the predicate matches, and one for when it doesn't. Then you would have full control over what happens in each case and we wouldn't need the `else` variants for each individual transform method, since you have two separate Unis.

Something like:

```java
uni.onItem().when("someValue"::equals)
	.ifTrue(trueUni -> trueUni.onItem().transform(this::doSomething))
	.ifFalse(falseUni -> falseUni.onItem().transform(this::doSomethingElse))
```

Not sure if that would be feasible and if it would actually be better then the current proposition. Main problem with this approach from my point of view would be that there should also be a way to "merge" the two Unis again, and I'm not sure whether there's a nice way to accomplish this.

Maybe one could also think of a syntax similar to `Uni.combine()`. Not sure. It's not a trivial problem :)

Reply:
Yes it's complicated